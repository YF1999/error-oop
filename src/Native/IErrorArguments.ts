import { IErrorOptions } from './Types';

export type IErrorOptionsKeys<T extends IErrorOptions> = ReadonlyArray<Extract<keyof T, string>>;

type PickRest<T extends IErrorOptions, List extends unknown[]> = List extends [unknown, ...infer R]
    ? R extends IErrorOptionsKeys<T>
        ? R
        : never
    : never;

type ArgsIteration<
    Prev extends unknown[],
    T extends IErrorOptions,
    Order extends IErrorOptionsKeys<T>,
> = Order extends [infer R, ...string[]]
    ? R extends keyof T
        ? [string, ...Prev, Required<T>[R], Error?] | ArgsIteration<[...Prev, Required<T>[R]], T, PickRest<T, Order>>
        : never
    : never;

export type IErrorArguments<T extends IErrorOptions, Order extends IErrorOptionsKeys<T> = []> =
    | []
    | [T]
    | [string, Error?]
    | ArgsIteration<[], T, Order>;
