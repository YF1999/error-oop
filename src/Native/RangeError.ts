import { ErrorOptions, RangeErrorMessageProps, RangeErrorProps } from '../CommonTypes';
import { setNonEnumerable } from '../utils';
import { appendInnerErrorStack } from './AppendInnerErrorStack';

export abstract class AbstractRangeError extends RangeError {
    #innerError?: Error;

    public constructor(props: RangeErrorProps, options: ErrorOptions<RangeErrorMessageProps>) {
        super();

        const { message, innerError } = props;
        const { generateMessage } = options;

        this.#innerError = innerError;
        this.message = generateMessage ? generateMessage({ name: this.name, message }) : message;

        // When the first call to `stack` property happens, it will combine `name` and `message` with trace stack to
        // `stack` property, we should generate message before this call.
        this.stack = appendInnerErrorStack(this.stack, this.#innerError);

        this._setNonEnumerable('message');
        this._setNonEnumerable('stack');
    }

    public get name(): string {
        return this.constructor.name;
    }

    public get innerError(): Error | undefined {
        return this.#innerError;
    }

    protected _setNonEnumerable(property: string): void {
        setNonEnumerable(this, property);
    }
}

/**
 * Represents an error that occurs when a numeric variable or parameter is outside of its valid range. This is roughly
 * the same as the native RangeError class. It additionally supports an innerError attribute.
 */
export class RangeErrorPro extends AbstractRangeError {
    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be append.
     */
    public constructor(message: string, innerError: Error);

    public constructor(message: string = '', innerError?: Error) {
        super({ message, innerError }, {});
    }
}

export { RangeErrorPro as RangeError };
