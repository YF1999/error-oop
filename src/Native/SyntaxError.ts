import { ErrorOptions, SyntaxErrorMessageProps, SyntaxErrorProps } from '../CommonTypes';
import { setNonEnumerable } from '../utils';
import { appendInnerErrorStack } from './AppendInnerErrorStack';

export abstract class AbstractSyntaxError extends SyntaxError {
    #innerError?: Error;

    public constructor(props: SyntaxErrorProps, options: ErrorOptions<SyntaxErrorMessageProps>) {
        super();

        const { message, innerError } = props;
        const { generateMessage } = options;

        this.#innerError = innerError;
        this.message = generateMessage ? generateMessage({ name: this.name, message }) : message;

        // When the first call to `stack` property happens, it will combine `name` and `message` with trace stack to
        // `stack` property, we should generate message before this call.
        this.stack = appendInnerErrorStack(this.stack, this.#innerError);

        this._setNonEnumerable('message');
        this._setNonEnumerable('stack');
    }

    public get name(): string {
        return this.constructor.name;
    }

    public get innerError(): Error | undefined {
        return this.#innerError;
    }

    protected _setNonEnumerable(property: string): void {
        setNonEnumerable(this, property);
    }
}

/**
 * Represents an error when trying to interpret syntactically invalid code. This is roughly the same as the native
 * SyntaxError class. It additionally supports an innerError attribute.
 */
export class SyntaxErrorPro extends AbstractSyntaxError {
    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be append.
     */
    public constructor(message: string, innerError: Error);

    public constructor(message: string = '', innerError?: Error) {
        super({ message, innerError }, {});
    }
}

export { SyntaxErrorPro as SyntaxError };
