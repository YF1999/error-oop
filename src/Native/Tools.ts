export function appendInnerErrorStack(currStack?: string, innerError?: Error): string | undefined {
    return innerError && innerError.stack
        ? `${currStack ?? 'No stack.'}\n--- inner error ---\n${innerError.stack}`
        : currStack;
}

export function setNonEnumerable(obj: object, property: string): void {
    Reflect.defineProperty(obj, property, {
        configurable: true,
        enumerable: false,
        writable: true,
    });
}
