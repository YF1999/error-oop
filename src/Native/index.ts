export * from './Error';
export * from './ErrorTool';
export * from './IErrorArguments';
export * from './RangeError';
export * from './ReferenceError';
export * from './SyntaxError';
export * from './TypeError';
export * from './Types';
export * from './URIError';
