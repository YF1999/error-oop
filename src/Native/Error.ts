import { appendInnerErrorStack, setNonEnumerable } from './Tools';
import { IErrorOptions, IStandardArguments } from './Types';

/**
 * This is roughly the same as the native Error class. It additionally supports an innerError attribute.
 */
export class NativeError extends Error {
    #innerError?: Error;

    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be appended.
     */
    public constructor(message: string, innerError: Error);
    /**
     * @param options The constructor options.
     */
    public constructor(options: IErrorOptions);

    public constructor(...[arg1, arg2]: IStandardArguments<IErrorOptions>) {
        super(typeof arg1 === 'object' ? arg1.message : arg1);

        // eslint-disable-next-line prefer-destructuring
        this.#innerError = typeof arg1 === 'object' ? arg1.innerError : arg2;

        // When the first call to `stack` property happens, it will combine `name` and `message` with trace stack to
        // `stack` property, we should generate message before this call.
        this.stack = appendInnerErrorStack(this.stack, this.#innerError);

        this._setNonEnumerable('stack');
    }

    public get name(): string {
        return this.constructor.name;
    }

    public get innerError(): Error | undefined {
        return this.#innerError;
    }

    protected _setNonEnumerable(property: string): void {
        setNonEnumerable(this, property);
    }
}

export { NativeError as Error };
