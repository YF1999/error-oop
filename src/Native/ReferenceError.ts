import { ReferenceErrorProps } from '../CommonTypes';
import { appendInnerErrorStack, setNonEnumerable } from './Tools';

export abstract class AbstractReferenceError extends ReferenceError {
    #innerError?: Error;

    public constructor(props: ReferenceErrorProps) {
        super(props.message);

        this.#innerError = props.innerError;

        // When the first call to `stack` property happens, it will combine `name` and `message` with trace stack to
        // `stack` property, we should generate message before this call.
        this.stack = appendInnerErrorStack(this.stack, this.#innerError);

        this._setNonEnumerable('stack');
    }

    public get name(): string {
        return this.constructor.name;
    }

    public get innerError(): Error | undefined {
        return this.#innerError;
    }

    protected _setNonEnumerable(property: string): void {
        setNonEnumerable(this, property);
    }
}

/**
 * Represents an error when a non-existent variable is referenced. This is roughly the same as the native ReferenceError
 * class. It additionally supports an innerError attribute.
 */
export class ReferenceErrorPro extends AbstractReferenceError {
    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be append.
     */
    public constructor(message: string, innerError: Error);

    public constructor(message: string = '', innerError?: Error) {
        super({ message, innerError });
    }
}

export { ReferenceErrorPro as ReferenceError };
