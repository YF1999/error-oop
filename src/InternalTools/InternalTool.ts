import { IErrorOptions, IErrorOptionsKeys } from '../Native';

export class InternalErrorTool {
    private static _parseErrorArguments<T extends IErrorOptions, Order extends IErrorOptionsKeys<T>>(
        order: Order,
        ..._args: [] | [T] | [string, ...unknown[]]
    ) {
        if (typeof _args[0] === 'object') {
            return _args[0];
        }

        const last = _args[_args.length - 1];
        const innerError = last instanceof Error ? (last as Error) : undefined;

        const args =
            innerError === undefined
                ? (_args as [string?, ...unknown[]])
                : (_args.slice(0, -1) as [string?, ...unknown[]]);

        const len = Math.min(order.length, args.length - 1);

        const options: IErrorOptions & Record<string, unknown> = { message: args[0], innerError };
        for (let i = 0; i < len; i++) {
            options[order[i]] = args[i + 1];
        }
        return options as T;
    }

    /**
     * message: string?[, arg1: T1?[, arg2: T2?[, ..., argn: Tn?]]], innerError: Error?
     */
    public static parseErrorArguments<T extends IErrorOptions, Order extends IErrorOptionsKeys<T>>(
        order: Order,
        ...args: [] | [T] | [string, ...unknown[]] // ...args: IErrorArguments<T, Order>
    ) {
        return InternalErrorTool._parseErrorArguments(order, ...args);
    }
}
