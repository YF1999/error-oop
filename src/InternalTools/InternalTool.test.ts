import { IErrorArguments, IErrorOptions } from '../Native';

import { InternalErrorTool } from './InternalTool';

interface IErrorObject extends IErrorOptions {
    a?: string;
    b?: number;
    c?: boolean;
    d?: string;
    e?: string;
}

describe('Test `InternalErrorTool`', () => {
    test('test `parseErrorArgument`', () => {
        const message = 'error object';
        const innerError = new Error();

        const args: IErrorArguments<IErrorObject, ['a', 'b', 'c', 'd', 'e']>[] = [
            [],
            [{ message }],
            [{ message, a: 'a', b: -1 }],
            [{ a: 'a', c: true, d: 'd', e: 'e' }],
            [message, 'a', -1, innerError],
            [message, 'a', -1, true],
            [message, 'a', -1, false, 'd', 'e', innerError],
        ];
        args.map((arg) =>
            expect(InternalErrorTool.parseErrorArguments(['a', 'b', 'c', 'd', 'e'], ...arg)).toMatchSnapshot(),
        );
    });
});
