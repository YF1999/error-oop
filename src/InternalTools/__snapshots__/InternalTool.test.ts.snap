// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test \`InternalErrorTool\` test \`parseErrorArgument\` 1`] = `
Object {
  "innerError": undefined,
  "message": undefined,
}
`;

exports[`Test \`InternalErrorTool\` test \`parseErrorArgument\` 2`] = `
Object {
  "message": "error object",
}
`;

exports[`Test \`InternalErrorTool\` test \`parseErrorArgument\` 3`] = `
Object {
  "a": "a",
  "b": -1,
  "message": "error object",
}
`;

exports[`Test \`InternalErrorTool\` test \`parseErrorArgument\` 4`] = `
Object {
  "a": "a",
  "c": true,
  "d": "d",
  "e": "e",
}
`;

exports[`Test \`InternalErrorTool\` test \`parseErrorArgument\` 5`] = `
Object {
  "a": "a",
  "b": -1,
  "innerError": [Error],
  "message": "error object",
}
`;

exports[`Test \`InternalErrorTool\` test \`parseErrorArgument\` 6`] = `
Object {
  "a": "a",
  "b": -1,
  "c": true,
  "innerError": undefined,
  "message": "error object",
}
`;

exports[`Test \`InternalErrorTool\` test \`parseErrorArgument\` 7`] = `
Object {
  "a": "a",
  "b": -1,
  "c": false,
  "d": "d",
  "e": "e",
  "innerError": [Error],
  "message": "error object",
}
`;
